File: 238.product-of-array-except-self.py
#
# @lc app=leetcode id=238 lang=python3
#
# [238] Product of Array Except Self
#

# @lc code=start
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        product = 1
        zero = False

        for num in nums:
            if num is not 0:
                product *= num
            elif zero is True:
                return [0] * len(nums)
            else:
                zero = True

        arr = [0] * len(nums)
        for i in range(len(nums)):
            if nums[i] is 0:
                arr[i] = product
            elif zero:
                arr[i] = 0
            else:
                arr[i] = product // nums[i]

        return arr

# @lc code=end

File: 238.product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int product = 1;
        boolean zero = false;

        for (int num : nums) {
            if (num == 0) {
                if (zero == true) {
                    return new int[nums.length];
                }
                zero = true;
            } else {
                product *= num;
            }
        }

        int[] ans = new int[nums.length];

        for (int i = 0; i < nums.length; i++) {
            int num = nums[i];

            if (num == 0) {
                ans[i] = product;
            } else if (zero) {
                ans[i] = 0;
            } else {
                ans[i] = product / num;
            }
        }

        return ans;
    }
}